airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,273 @@
+from pyrate_limiter import Rate as OrgRate
+from pyrate_limiter.exceptions import BucketFullException
+from requests_cache import CachedSession
+
+Duration = OrgDuration
+Rate = OrgRate
+
+
+class CallRateLimitHit(Exception):
+    def __init__(self, error: str, item: Any, weight: int, rate: str, time_to_wait: timedelta):
+        """Constructor
+
+        :param error: error message
+        :param item: object passed into acquire_call
+        :param weight: how many credits were requested
+        :param rate: string representation of the rate violated
+        :param time_to_wait: how long should wait util more call will be available
+        """
+        self.item = item
+        self.weight = weight
+        self.rate = rate
+        self.time_to_wait = time_to_wait
+        super().__init__(error)
+
+
+class AbstractCallRatePolicy(abc.ABC):
+    """Call rate policy interface.
+    Should be configurable with different rules, like N per M for endpoint X
+    """
+
+    @abc.abstractmethod
+    def try_acquire(self, request: Any, weight: int) -> None:
+        """Try to acquire request
+
+        :param request: request object representing single call to API
+        :param weight: number of requests to deduct from credit
+        :return:
+        """
+
+
+class CallRatePolicy(AbstractCallRatePolicy):
+    """
+    Policy to control requests rate implemented on top of PyRateLimiter lib.
+
+    TODO: periodical clean up of the bucket
+    TODO: support static window strategy, not only moving window
+    TODO: support policy without limitations
+    """
+
+    def __init__(self, rates: list[Rate]):
+        """Constructor
+
+        :param rates: list of rates, the order is important and must be ascending
+        """
+        self._bucket = InMemoryBucket(rates)
+        # Limiter will create background task that clears old requests in the bucket
+        self._limiter = Limiter(self._bucket)
+
+    def try_acquire(self, request: Any, weight: int = 1) -> None:
+        try:
+            self._limiter.try_acquire(request, weight=weight)
+        except BucketFullException as exc:
+            item = self._limiter.bucket_factory.wrap_item(request, weight)
+            time_to_wait: int = self._bucket.waiting(item)  # type: ignore
+            raise CallRateLimitHit(
+                error=str(exc.meta_info["error"]),
+                item=request,
+                weight=int(exc.meta_info["weight"]),
+                rate=str(exc.meta_info["rate"]),
+                time_to_wait=timedelta(milliseconds=time_to_wait),
+            )
+
+
+class RequestMatcher(abc.ABC):
+    """Callable that help to match request object with call rate policies."""
+
+    @abc.abstractmethod
+    def __call__(self, request: Any) -> bool:
+        """
+
+        :param request:
+        :return: True if pattern matches the provided request object, False - otherwise
+        """
+
+
+class HttpRequestMatcher(RequestMatcher):
```

Comment:
Why do we need this concept? Can't we just instantiate different `CallRatePolicy`? It feels like our HttpStreams do not query on different paths and I don't know of an API that rate limits differently based on query params or headers

Reply:
1. I had to deal with at least a few APIs with different call rate limits for different resources. I will try to lookup some as an example. 
2. There is no restriction for HttpStream to call only a single endpoint, I can't prove or disprove that we have such streams at the moment, but I can think about cases when HttpStream need to do other queries to API.
3. Simplification of the interface will make it really hard to go generic and work with rates beyong Http streams and sources in general.

Reply:
> I had to deal with at least a few APIs with different call rate limits for different resources. I will try to lookup some as an example.

Are these in different streams using different HttpStream?

> Simplification of the interface will make it really hard to go generic and work with rates beyong Http streams and sources in general.

The other way around is true i.e. having more code is more code to support. 

Just to be clear, I'm not saying this isn't a good thing. My only concern is that this is code that won't be leveraged in prod and the we will assume that it's working until it doesn't and it'll be hard to traceback to this change.

Reply:
> Are these in different streams using different HttpStream?

I mean HttpStream using requests.session directly for other endpoints

> The other way around is true i.e. having more code is more code to support.
Agree, and I would love to have a simple clear interface. But at the same time extending interfaces also challenging after they got into production and now used everywhere. There is no strict requirement to use HttpMatcher for simple cases like one Policy for all endpoints, or one endpoint for one HttpStream. And for such scenarios, we can always have a default matcher returning True. I see it as a small cost/overhead for having greater flexibility in future. WDYT?

Reply:
@maxi297 FYI, an example of API with different rates per endpoints - https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limiting